pkg/sources/forward/data_forward.go

```diff
@@ -0,0 +1,621 @@
+func (isdf *DataForward) forwardAChunk(ctx context.Context) {
+	start := time.Now()
+	// There is a chance that we have read the message and the container got forcefully terminated before processing. To provide
+	// at-least-once semantics for reading, during restart we will have to reprocess all unacknowledged messages. It is the
+	// responsibility of the Read function to do that.
+	readMessages, err := isdf.reader.Read(ctx, isdf.opts.readBatchSize)
+	if err != nil {
+		isdf.opts.logger.Warnw("failed to read from source", zap.Error(err))
+		forward.ReadMessagesError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelPartitionName: isdf.reader.GetName()}).Inc()
+	}
+	forward.ReadMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelPartitionName: isdf.reader.GetName()}).Add(float64(len(readMessages)))
+
+	// process only if we have any read messages. There is a natural looping here if there is an internal error while
+	// reading, and we are not able to proceed.
+	if len(readMessages) == 0 {
+		// When the read length is zero, the write length is definitely zero too
+		// meaning there's no data to be published to the next vertex, and we consider this
+		// situation as idling.
+		// In order to continue propagating watermark, we will set watermark idle=true and publish it.
+		// We also publish a control message if this is the first time we get this idle situation.
+		// we use the HeadWMB as the watermark for the idle
+		// we get the HeadWMB for the partition from which we read the messages
+		var processorWMB = isdf.wmFetcher.GetHeadWMB(isdf.reader.GetPartitionIdx())
+		if !isdf.wmbChecker.ValidateHeadWMB(processorWMB) {
+			// validation failed, skip publishing
+			isdf.opts.logger.Debugw("skip publishing idle watermark",
+				zap.Int("counter", isdf.wmbChecker.GetCounter()),
+				zap.Int64("offset", processorWMB.Offset),
+				zap.Int64("watermark", processorWMB.Watermark),
+				zap.Bool("Idle", processorWMB.Idle))
+			return
+		}
+
+		// if the validation passed, we will publish the watermark to all the toBuffer partitions.
+		for toVertexName, toVertexBuffer := range isdf.toBuffers {
+			for _, partition := range toVertexBuffer {
+				if p, ok := isdf.toVertexWMPublishers[toVertexName]; ok {
+					idlehandler.PublishIdleWatermark(ctx, partition, p, isdf.idleManager, isdf.opts.logger, dfv1.VertexTypeSource, wmb.Watermark(time.UnixMilli(processorWMB.Watermark)))
+				}
+			}
+		}
+		return
+	}
+
+	var dataMessages = make([]*isb.ReadMessage, 0, len(readMessages))
+	// store the offsets of the messages we read from source
+	var readOffsets = make([]isb.Offset, len(readMessages))
+	for idx, m := range readMessages {
+		readOffsets[idx] = m.ReadOffset
+		// for source forwarder, below if statement should always hold true, since we never publish idle watermark to source reader buffer
```

Comment:
@jy4096 Hey Juanlu, I am not quite familiar with idle watermark. Could you help identify if I could remove more code blocks regarding idle watermark for this source forwarder? Right now I am thinking maybe I can remove the concept of `dataMessages` from forwarder and just use `readMessages` instead. Because we never send ctrl messages to source.